src/core/utils/useSummarizeChat.ts

```diff
@@ -0,0 +1,44 @@
+import { useState, useEffect } from 'react';
+import OpenAI from 'openai';
+import { ChatCompletionMessageParam } from 'openai/resources/index';
+
+export const useSummarizeChat = (
+  client: OpenAI,
+  messages: ChatCompletionMessageParam[],
+): string => {
+  const [summary, setSummary] = useState('');
+
+  useEffect(() => {
+    if (!messages?.length) return;
+
+    const generateSummary = async () => {
+      const initialMessages = messages.slice(0, 3);
+
+      try {
+        const completion = await client.chat.completions.create({
+          model: 'gpt-4o-mini',
+          messages: [
+            {
+              role: 'system',
+              content:
+                'Generate a brief, descriptive title (max 60 chars) for this chat based on the first user message.',
+            },
+            ...initialMessages,
+          ],
+        });
+
+        setSummary(completion.choices[0].message.content as string);
+      } catch (error) {
+        console.error('Error generating summary:', error);
+        const firstMessage = initialMessages.find(
+          (m) => m.role === 'user',
+        )?.content;
+        setSummary(firstMessage?.slice(0, 57) + '...' || 'New Chat');
+      }
+    };
+
+    generateSummary();
+  }, [messages, client]);
```

Comment:
one problem with this is that the messages array's reference changes completely everytime we add a new message entry to it. so as soon as the messages array has some length greater than 1 it will call fetch and try to summarize everytime we add a new message.

Reply:
Sure - feel free to adjust this code or start from scratch. Just meant as a template/POC